topic-operator/src/main/java/io/strimzi/operator/topic/Session.java

```diff
@@ -133,72 +136,102 @@ public void init(Vertx vertx, Context context) {
      * Stop the operator.
      */
     @Override
-    @SuppressWarnings("deprecation") // Uses a deprecated executeBlocking call that should be addressed later. This is tracked in https://github.com/strimzi/strimzi-kafka-operator/issues/9233
     public void stop(Promise<Void> stop) throws Exception {
+        LOGGER.info("Stopping");
         this.stopped = true;
-        Long timerId = this.timerId;
-        if (timerId != null) {
-            vertx.cancelTimer(timerId);
+        if (this.timerId != null) {
+            vertx.cancelTimer(this.timerId);
         }
-        vertx.executeBlocking(blockingResult -> {
-            long timeout = 120_000L;
-            long deadline = System.currentTimeMillis() + timeout;
-            LOGGER.info("Stopping");
-            LOGGER.debug("Stopping kube watch");
-            topicWatch.close();
-            LOGGER.debug("Stopping zk watches");
-            topicsWatcher.stop();
-
-            Promise<Void> promise = Promise.promise();
-            Handler<Long> longHandler = new Handler<>() {
-                @Override
-                public void handle(Long inflightTimerId) {
-                    if (!topicOperator.isWorkInflight()) {
-                        LOGGER.debug("Inflight work has finished");
-                        promise.complete();
-                    } else if (System.currentTimeMillis() > deadline) {
-                        LOGGER.error("Timeout waiting for inflight work to finish");
-                        promise.complete();
-                    } else {
-                        LOGGER.debug("Waiting for inflight work to finish");
-                        vertx.setTimer(1_000, this);
+        ExecutorService shutdownExecutor = Executors.newSingleThreadExecutor();
+        closeKubernetesWatch()
+                .onSuccess(ignored -> LOGGER.debug("Kubernetes watch closed"))
+                .onFailure(cause -> LOGGER.error("Failed to close Kubernetes watch", cause))
+            .compose(ignored -> stopZooKeeperWatcher())
```

Comment:
I think we could close both kube and ZK watches in parallel. The benefit of doing that is it leaves more time (out of the total 10s hard limit applied by the shutdown hook) to cleanly shutdown the topic store. Note that we don't actually care too much about waiting for things like underlying TCP connections to close. All we really need is that new events stop being delivered to the operator so it can work through in flight events prior to the closing of the topic store.